enterprise/server/remote_execution/cgroup/cgroup.go

```diff
@@ -0,0 +1,189 @@
+		cpuNanos, err = readInt64FromFile(cpuUsagePath)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// cgroup v2: /cpu.stat file contains a line like "usage_usec <N>" It
+		// contains other lines like user_usec, system_usec etc. but we just
+		// report the total for now.
+		cpuMicros, err := readCgroupInt64Field(cpuUsagePath, "usage_usec")
+		if err != nil {
+			return nil, err
+		}
+		cpuNanos = cpuMicros * 1e3
+	}
+	return &repb.UsageStats{
+		MemoryBytes: memUsageBytes,
+		CpuNanos:    cpuNanos,
+	}, nil
+}
+
+func (p *Paths) CgroupVersion() int {
+	if p.CPUTemplate == "" || p.MemoryTemplate == "" {
+		// Not fully initialized.
+		return 0
+	} else if strings.HasSuffix(p.CPUTemplate, "/cpuacct.usage") {
+		return 1
+	} else {
+		return 2
+	}
+}
+
+// find locates cgroup path templates. For this to work, the container must be
+// started (i.e. `podman create` does not set up the cgroups; `podman start`
+// does). We use this walking approach because the logic for figuring out the
+// actual cgroup paths depends on the system setup and is pretty complicated.
+func (p *Paths) find(ctx context.Context, cid string) error {
+	// If already initialized, do nothing.
+	p.mu.RLock()
+	v := p.CgroupVersion()
+	p.mu.RUnlock()
+	if v != 0 {
+		return nil
+	}
+
+	// Walk the cgroupfs tree.
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	// Re-check whether we're initialized since we released the lock for a brief
+	// period.
+	if p.CgroupVersion() != 0 {
+		return nil
+	}
+	start := time.Now()
+	// Sentinel error value to short-circuit the walk
+	stop := fmt.Errorf("stop walk")
+	err := filepath.WalkDir(cgroupfsPath, func(path string, dir fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if !strings.Contains(path, cid) {
+			return nil
+		}
+		if strings.HasSuffix(path, "/memory.usage_in_bytes") || strings.HasSuffix(path, "/memory.current") {
+			p.MemoryTemplate = strings.ReplaceAll(path, cid, cidPlaceholder)
+		} else if strings.HasSuffix(path, "/cpu.stat") || strings.HasSuffix(path, "/cpuacct.usage") {
+			p.CPUTemplate = strings.ReplaceAll(path, cid, cidPlaceholder)
+		}
+		if p.MemoryTemplate != "" && p.CPUTemplate != "" {
+			return stop
+		}
+		return nil
+	})
+	if err != nil && err != stop {
+		return err
+	}
+	if p.MemoryTemplate == "" {
+		return status.InternalErrorf("failed to locate memory cgroup file under %s", cgroupfsPath)
+	}
+	if p.CPUTemplate == "" {
+		return status.InternalErrorf("failed to locate CPU cgroup file under %s", cgroupfsPath)
+	}
+	log.CtxInfof(ctx, "Initialized cgroup path templates (%s): mem=%s, cpu=%s", time.Since(start), p.MemoryTemplate, p.CPUTemplate)
+	return nil
+}
+
+// readInt64FromFile reads a file expected to contain a single int64.
+func readInt64FromFile(path string) (int64, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return 0, err
+	}
+	defer f.Close()
+	b, err := io.ReadAll(f)
+	if err != nil {
+		return 0, err
+	}
```

Comment:
can you use `data, err := os.ReadFile(path)` here?

Reply:
yep done